compose/foundation/foundation/src/desktopMain/kotlin/androidx/compose/foundation/MouseCombinedClickable.desktop.kt

```diff
@@ -0,0 +1,543 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package androidx.compose.foundation
+
+import androidx.compose.foundation.gestures.GestureCancellationException
+import androidx.compose.foundation.gestures.PressGestureScope
+import androidx.compose.foundation.interaction.MutableInteractionSource
+import androidx.compose.foundation.interaction.PressInteraction
+import androidx.compose.runtime.MutableState
+import androidx.compose.runtime.State
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.remember
+import androidx.compose.runtime.rememberUpdatedState
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.composed
+import androidx.compose.ui.geometry.Offset
+import androidx.compose.ui.geometry.Size
+import androidx.compose.ui.input.key.KeyEvent
+import androidx.compose.ui.input.key.isAltPressed
+import androidx.compose.ui.input.key.isCtrlPressed
+import androidx.compose.ui.input.key.isMetaPressed
+import androidx.compose.ui.input.key.isShiftPressed
+import androidx.compose.ui.input.key.onKeyEvent
+import androidx.compose.ui.input.pointer.AwaitPointerEventScope
+import androidx.compose.ui.input.pointer.PointerButtons
+import androidx.compose.ui.input.pointer.PointerEvent
+import androidx.compose.ui.input.pointer.PointerEventPass
+import androidx.compose.ui.input.pointer.PointerInputChange
+import androidx.compose.ui.input.pointer.PointerInputScope
+import androidx.compose.ui.input.pointer.PointerKeyboardModifiers
+import androidx.compose.ui.input.pointer.PointerType
+import androidx.compose.ui.input.pointer.changedToDown
+import androidx.compose.ui.input.pointer.changedToUp
+import androidx.compose.ui.input.pointer.isOutOfBounds
+import androidx.compose.ui.input.pointer.isPrimaryPressed
+import androidx.compose.ui.input.pointer.pointerInput
+import androidx.compose.ui.platform.debugInspectorInfo
+import androidx.compose.ui.semantics.Role
+import androidx.compose.ui.semantics.disabled
+import androidx.compose.ui.semantics.onClick
+import androidx.compose.ui.semantics.onLongClick
+import androidx.compose.ui.semantics.role
+import androidx.compose.ui.semantics.semantics
+import androidx.compose.ui.unit.Density
+import androidx.compose.ui.util.fastAll
+import androidx.compose.ui.util.fastAny
+import androidx.compose.ui.util.fastForEach
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.coroutineScope
+import kotlinx.coroutines.currentCoroutineContext
+import kotlinx.coroutines.isActive
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.sync.Mutex
+
+@ExperimentalFoundationApi
+fun Modifier.onPrimaryCombinedClickable(
```

Comment:
@igordmn here is a variant how we can handle all pointer types (mouse, touch, stylus) for primary clicks/longClicks/doubleClicks + filter by PointerKeyboardModifiers. 
 In case when it's mouse it reacts only when `isPrimaryPressed` becomes true, and for touch/stylus it reacts on `changedToDown`.

`Modifier. combinedMouseClickable` still exists to handle mouse clicks with specific `PointerButtons`.

Both modifiers will be available in common code, so something like this will be possible:

```kotlin
Modifier.onPrimaryCombinedClickable(
    enabled = enabled,
    onDoubleClick = {
        println("Simple LClick DoubleClick")
    },
    onLongPress = {
        println("Simple LClick LongPress")
    }) {
        println("Simple LClick click")
    }
    .combinedMouseClickable(
        enabled = enabled,
        buttons = { it.isSecondaryPressed },
        keyModifiers = { it.isShiftPressed } ) 
   {
       println("RClick + Shit")
   }
```

Obviously, combinedMouseClickable won't react on touch events. 

We also may add `onSecondaryCombinedClickable`. But we need to decide what touch events should correspond to isSecondaryPressed for mouse. Another option (probably better) is maybe making `Modifier.customCombinedClickable ` public, so anyone can create their custom semantic event handlers relevant to their apps by passing proper `pressFilter ` and `releaseFilter` (with access to PointerEvent).

What do you think? It's a big comment, so we can discuss step by step.

Reply:
It would be wonderful, if we have `customCombinedClickable` public. I have two concerns:

1. Maybe because we have multiple names (combinedMouseClickable, combinedMouseClickable, onPrimaryCombinedClickable, customCombinedClickable), users may be confused which function they should use. If we would have only one name `combinedClickable`, they could just know about this function, and investigate further its features, looking at the parameter list. If there is no parameter specified, by default we detect a primary click.
2. Having pressFilter/releaseFilter is verbose, and it's rarely needed to provide non-inverted filters. What if we introduce `PointerButton` class? With this class we can merge `pressFilter`, `releaseFilter`, `buttons`, `keyboardModifiers` into one filter. We can check this filter on every press/release event and on every `PointerInputChange`. To detect if it was a press or a release, we probably can compare every PointerButtons state and every PointerInputChange.changedToDown/changedTopUp. I still don't know if it is possible, but if it is, it would simplify API.

```
combinedClickable(
    filter = { type == PointerType.Mouse && button == PointerButton.Secondary }
    onDoubleClick={}
)

combinedClickable(
    filter = { type == PointerType.Mouse && button == PointerButton.Secondary && keyboardModifiers.isCtrlPressed && keyboardModifiers.isShiftPressed }
    onDoubleClick={}
)

combinedClickable(
  // filter = { isPrimary } // default value
  onDoubleClick={}
)

combinedClickable(
    filter = { isPrimary && keyboardModifiers.isCtrlPressed }
    onDoubleClick={}
)

class PointerFilterScope(
    val type: PointerType
    val button: PointerButton
    val keyboardModifiers: PointerKeyboardModifiers
)

val PointerFilterScope.isPrimary get() = type == PointerType.Mouse && button == PointerButton.Primary || type == PointerType.Touch
```

> We also may add onSecondaryCombinedClickable. But we need to decide what touch events should correspond to isSecondaryPressed for mouse

As we discussed before, it is hard to decide what is a secondary click for touch. Maybe we can have something less abstract, similar to `java.awt.event.MouseEvent.isPopupTrigger`:
```
Modifier.onContextMenuTrigger { } // press+release for mouse on Windows/Linux, press for mouse on macOs, long press for touch
```
But we can keep this functionality out of scope of the current work.

Reply:
>  introduce PointerButton

In this case we can encounter the same issue as with checking PointerEvenType.Press instead of the state of the buttons 🤔. For example, we can miss release event for button2, and we will stuck in click state.

Maybe I was not right about the fact that we should check button state instead of pointerEventType, and it is an issue of the platform, not of our implementation. We can discuss it online.